aws/resource_aws_pinpoint_event_stream.go

```diff
@@ -42,19 +44,22 @@ func resourceAwsPinpointEventStreamUpsert(d *schema.ResourceData, meta interface
 
 	applicationId := d.Get("application_id").(string)
 
-	params := &pinpoint.WriteEventStream{}
-
-	params.DestinationStreamArn = aws.String(d.Get("destination_stream_arn").(string))
-	params.RoleArn = aws.String(d.Get("role_arn").(string))
+	params := &pinpoint.WriteEventStream{
+		DestinationStreamArn: aws.String(d.Get("destination_stream_arn").(string)),
+		RoleArn:              aws.String(d.Get("role_arn").(string)),
+	}
 
 	req := pinpoint.PutEventStreamInput{
 		ApplicationId:    aws.String(applicationId),
 		WriteEventStream: params,
 	}
 
-	_, err := conn.PutEventStream(&req)
+	// Retry for IAM eventual consistency
+	_, err := retryOnAwsCode("BadRequestException", func() (interface{}, error) {
+		return conn.PutEventStream(&req)
+	})
```

Comment:
IAM eventual consistency retry logic should follow this Contributing Guide section: https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/retries-and-waiters.md#iam-error-retries

I would also recommend conditionalizing this based on the error message as well, since `BadRequestException` by itself could retry on actually malformed requests.